apps/ejabberd/src/mod_offline_riak.erl

```diff
@@ -0,0 +1,187 @@
+    {ok, length(Keys)}.
+
+-spec remove_user(LUser, LServer) -> any() when
+    LUser :: binary(),
+    LServer :: binary().
+
+remove_user(LUser, LServer) ->
+    Keys = read_user_idx(LUser, LServer),
+    [mongoose_riak:delete(bucket_type(LServer), Key) || Key <- Keys].
+
+read_user_idx(LUser, LServer) ->
+    {ok, IdxResult} = mongoose_riak:get_index(bucket_type(LServer), ?USER_IDX,
+                                              LUser, []),
+    IdxResult?INDEX_RESULTS.keys.
+
+write_msg(LUser, LServer, #offline_msg{from = FromJID, packet = Packet,
+                                       timestamp = TimestampIn, expire = Expire}) ->
+    Timestamp = usec:from_now(TimestampIn),
+    Obj = riakc_obj:new(bucket_type(LServer), key(LUser, Timestamp), exml:to_binary(Packet)),
+    MD = riakc_obj:get_update_metadata(Obj),
+    SecondaryIndexes = [{?TIMESTAMP_IDX, [Timestamp]},
+                        {?USER_IDX, [LUser]},
+                        {?EXPIRE_IDX, [maybe_encode_timestamp(Expire)]}],
+    MDWithIndexes = riakc_obj:set_secondary_index(MD, SecondaryIndexes),
+    From = jid:to_binary(FromJID),
+    UserMetaData = [{<<"from">>, From}],
+    MDWithUserData = set_obj_user_metadata(MDWithIndexes, UserMetaData),
+    FinalObj = riakc_obj:update_metadata(Obj, MDWithUserData),
+    mongoose_riak:put(FinalObj).
+
+set_obj_user_metadata(MD, []) ->
+    MD;
+set_obj_user_metadata(MD, [UserMD | Rest]) ->
+    UpdatedMD = riakc_obj:set_user_metadata_entry(MD, UserMD),
+    set_obj_user_metadata(UpdatedMD, Rest).
+
+pop_msg(Key, LUser, LServer, To) ->
+    try
+        {ok, Obj} = mongoose_riak:get(bucket_type(LServer), Key),
+
+        PacketRaw = riakc_obj:get_value(Obj),
+        {ok, Packet} = exml:parse(PacketRaw),
+        MD = riakc_obj:get_update_metadata(Obj),
+        [Timestamp] = riakc_obj:get_secondary_index(MD, ?TIMESTAMP_IDX),
+        [Expire] = riakc_obj:get_secondary_index(MD, ?EXPIRE_IDX),
+        User = riakc_obj:get_secondary_index(MD, ?USER_IDX),
+        From = riakc_obj:get_user_metadata_entry(MD, <<"from">>),
+
+        mongoose_riak:delete(bucket_type(LServer), Key),
+
+        #offline_msg{us = {LUser, LServer},
+                     timestamp = usec:to_now(Timestamp),
+                     expire = maybe_decode_timestamp(Expire),
+                     from = jid:from_binary(From),
+                     to = To,
+                     packet = Packet}
+
+    catch
+        Error:Reason ->
+            ?WARNING_MSG("issue=~p, action=reading_key, host=~s, reason=~p, stack_trace=~p",
+                         [Error, LServer, Reason, erlang:get_stacktrace()]),
+            []
+    end.
+
+
+-spec bucket_type(ejabberd:lserver()) -> {binary(), ejabberd:lserver()}.
+bucket_type(LServer) ->
+    {<<"offline">>, LServer}.
+
+-spec key(binary(), integer()) -> binary().
+key(LUser, TimestampInt) ->
+    Timestamp = integer_to_binary(TimestampInt),
+    Random = integer_to_binary(random:uniform(1024)),
```

Comment:
I added the `random:seed` in `mod_offline:init` function as this function will be called from that process.